src/pluginextension.h

```diff
@@ -0,0 +1,76 @@
+/*
+ *      pluginextension.h - this file is part of Geany, a fast and lightweight IDE
+ *
+ *      Copyright 2023 The Geany contributors
+ *
+ *      This program is free software; you can redistribute it and/or modify
+ *      it under the terms of the GNU General Public License as published by
+ *      the Free Software Foundation; either version 2 of the License, or
+ *      (at your option) any later version.
+ *
+ *      This program is distributed in the hope that it will be useful,
+ *      but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *      GNU General Public License for more details.
+ *
+ *      You should have received a copy of the GNU General Public License along
+ *      with this program; if not, write to the Free Software Foundation, Inc.,
+ *      51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#ifndef GEANY_PLUGIN_EXTENSION_H
+#define GEANY_PLUGIN_EXTENSION_H 1
+
+#include "document.h"
+
+/* Temporary define so the plugin can check whether it's compiled against
+ * Geany with plugin extension support or not */
+#define HAVE_GEANY_PLUGIN_EXTENSION 1
+
+G_BEGIN_DECLS
+
+
+typedef struct {
+	gboolean (*autocomplete_provided)(GeanyDocument *doc);
+	void (*autocomplete_perform)(GeanyDocument *doc, gboolean force);
+
+	gboolean (*calltips_provided)(GeanyDocument *doc);
+	void (*calltips_show)(GeanyDocument *doc, gboolean force);
+
+	gboolean (*goto_provided)(GeanyDocument *doc);
+	void (*goto_perform)(GeanyDocument *doc, gint pos, gboolean definition);
+
+	gboolean (*doc_symbols_provided)(GeanyDocument *doc);
+	GPtrArray *(*doc_symbols_get)(GeanyDocument *doc);
+
+	gboolean (*symbol_highlight_provided)(GeanyDocument *doc);
```

Comment:
I guess there isn't any `_perform()` because there isn't a canonical place where everybody should do this?

Reply:
Basically there's no need for any `perform()` here. In the other API calls it is necessary because the plugin needs to know when e.g. a keybinding invoking goto definition/declaration was pressed to perform the corresponding action. Here, instead, the plugin can perform the highlighting by itself e.g. when the current document becomes visible or the user types something and the only thing necessary is that geany knows highlighting is provided by the plugin so it doesn't do anything.